File: cracking_the_coding_interview_6th/chapter02/python/2_03.py
# https://www.geeksforgeeks.org/data-structures/linked-list/
class Node:
    def __init__(self):
        self.data: int = -1
        self.nxt: Node = None


class LinkedList:
    def __init__(self):
        self.head: Node = None

    def __repr__(self):
        node = self.head
        lst = []
        while node is not None:
            lst.append(node.data)
            node = node.nxt
        return repr(lst)

    def __len__(self):
        cnt = 0
        cur = self.head
        while cur is not None:
            cur = cur.nxt
            cnt += 1
        return cnt

    def push(self, new_data: int):
        new_node = Node()
        new_node.data = new_data
        new_node.nxt = self.head
        self.head = new_node

    def insert_after(self, prev_node: Node, new_data: int):
        assert prev_node is not None

        new_node = Node()
        new_node.data = new_data
        new_node.nxt = prev_node.nxt
        prev_node.nxt = new_node

    def append(self, new_data: int):
        new_node = Node()
        new_node.data = new_data
        new_node.nxt = None

        if self.head is None:
            self.head = new_node
            return

        last = self.head
        while last.nxt is not None:
            last = last.nxt
        last.nxt = new_node

    def delete_node(self, key):
        tmp = self.head
        prv = None

        if tmp is not None and tmp.data == key:
            self.head = tmp.nxt
            return

        while tmp is not None and tmp.data != key:
            prv = tmp
            tmp = tmp.nxt

        if tmp is None:
            return

        prv.nxt = tmp.nxt

    def delete_node_by_pos(self, pos):
        if self.head is None:
            return

        tmp = self.head

        if pos == 0:
            self.head = tmp.nxt
            return

        for _ in range(pos - 1):
            tmp = tmp.nxt
            if tmp is None:
                return

        if tmp.nxt is None:
            return

        tmp.nxt = tmp.nxt.nxt

    def delete_list(self):
        self.head = None

    def search(self, x):
        cur = self.head
        while cur is not None:
            if cur.data == x:
                return True
            cur = cur.nxt
        return False

    def get(self, pos):
        cur = self.head
        cnt = 0
        while cur is not None:
            if cnt == pos:
                return cur.data
            cnt += 1
            cur = cur.nxt
        raise IndexError("index out of range")

    def get_ptr(self, key):
        cur = self.head
        while cur is not None:
            if cur.data == key:
                return cur
            cur = cur.nxt
        return cur


def del_md_node1(md: Node):
    if md is None or md.nxt is None:
        return False
    md.data = md.nxt.data
    md.nxt = md.nxt.nxt
    return True


def make_list(s):
    lst = LinkedList()
    for c in s:
        lst.append(ord(c))
    return lst


def run(f, fn, ps):
    print(f"# {fn}")
    for k, s in ps:
        k = ord(k)
        lst = make_list(s)
        before = repr(lst)
        f(lst.get_ptr(k))
        after = repr(lst)
        print(f"{k}: {before} -> {after}")
    print()


def ns(f):
    return next(f).strip()


with open("../testcases/2_03.txt") as f:
    n = int(ns(f))
    ps = [ns(f).split() for _ in range(n)]

run(del_md_node1, "del_md_node1", ps)

File: cracking_the_coding_interview_6th/chapter02/cpp/2_03.cpp
// https://www.geeksforgeeks.org/data-structures/linked-list/
#include <cassert>
#include <fstream>
#include <functional>
#include <string>
#include <utility>
#include <vector>
using namespace std;
typedef long long ll;
typedef pair<ll, string> pls;

struct Node {
  ll data;
  Node *next;
};

struct LinkedList {
  Node *head = nullptr;

  string to_str() {
    Node *node = head;
    string res = "[";
    while (node != nullptr) {
      res += to_string(node->data);
      res += " ";
      node = node->next;
    }
    if (res.size() > 1) res.pop_back();
    res += "]";
    return res;
  }

  void push(ll new_data) {
    Node *new_node = new Node();
    new_node->data = new_data;
    new_node->next = head;
    head = new_node;
  }

  void insert_after(Node *prev_node, ll new_data) {
    assert(prev_node != nullptr);

    Node *new_node = new Node();
    new_node->data = new_data;
    new_node->next = prev_node->next;
    prev_node->next = new_node;
  }

  void append(ll new_data) {
    Node *new_node = new Node();
    new_node->data = new_data;
    new_node->next = nullptr;

    if (head == nullptr) {
      head = new_node;
      return;
    }

    Node *last = head;
    while (last->next != nullptr) {
      last = last->next;
    }
    last->next = new_node;
  }

  void delete_node(ll key) {
    Node *tmp = head;
    Node *prev = nullptr;

    // head has the key
    if (tmp != nullptr && tmp->data == key) {
      head = tmp->next;
      delete tmp;
      return;
    }

    while (tmp != nullptr && tmp->data != key) {
      prev = tmp;
      tmp = tmp->next;
    }

    // not found a node having the key
    if (tmp == nullptr) return;

    // found
    prev->next = tmp->next;
    delete tmp;
  }

  void delete_node_by_pos(ll pos) {
    if (head == nullptr) return;

    Node *tmp = head;

    if (pos == 0) {
      head = tmp->next;
      delete tmp;
      return;
    }

    for (ll _ = 0; _ < pos - 1; ++_) {
      tmp = tmp->next;
      if (tmp == nullptr) return;
    }

    if (tmp->next == nullptr) return;

    Node *node = tmp->next;
    tmp->next = node->next;
    delete node;
  }

  void delete_list() {
    Node *cur = head;
    Node *tmp = nullptr;

    while (cur != nullptr) {
      tmp = cur->next;
      delete cur;
      cur = tmp;
    }

    head = nullptr;
  }

  ll length() {
    ll cnt = 0;
    Node *cur = head;
    while (cur != nullptr) {
      cur = cur->next;
      ++cnt;
    }
    return cnt;
  }

  bool search(ll x) {
    Node *cur = head;
    while (cur != nullptr) {
      if (cur->data == x) return true;
      cur = cur->next;
    }
    return false;
  }

  ll get(ll pos) {
    Node *cur = head;
    ll cnt = 0;
    while (cur != nullptr) {
      if (cnt == pos) return cur->data;
      ++cnt;
      cur = cur->next;
    }
    throw "Index out of range";
  }

  Node *get_ptr(ll key) {
    Node *cur = head;
    while (cur != nullptr) {
      if (cur->data == key) return cur;
      cur = cur->next;
    }
    return cur;
  }
};

LinkedList make_list(string s) {
  LinkedList lst;
  for (auto c : s)
    lst.append(c);
  return lst;
}

void del_md_node1(Node *md) {
  md->data = md->next->data;
  Node *tmp = md->next;
  md->next = md->next->next;
  delete tmp;
}

bool del_md_node2(Node *md) {
  if (md == nullptr || md->next == nullptr) return false;
  Node *tmp = md->next;
  md->data = md->next->data;
  md->next = md->next->next;
  delete tmp;
  return true;
}

void run1(const function<void(Node *)> &f,
          const string &fn,
          const vector<pls> &ps) {
  printf("# %s\n", fn.c_str());
  for (auto p : ps) {
    ll k = p.first;
    string s = p.second;
    LinkedList lst = make_list(s);
    string before = lst.to_str();
    f(lst.get_ptr(k));
    string after = lst.to_str();
    printf("%lld: %s -> %s\n", k, before.c_str(), after.c_str());
  }
  puts("");
}

void run2(const function<bool(Node *)> &f,
          const string &fn,
          const vector<pls> &ps) {
  printf("# %s\n", fn.c_str());
  for (auto p : ps) {
    ll k = p.first;
    string s = p.second;
    LinkedList lst = make_list(s);
    string before = lst.to_str();
    f(lst.get_ptr(k));
    string after = lst.to_str();
    printf("%lld: %s -> %s\n", k, before.c_str(), after.c_str());
  }
  puts("");
}

int main() {
  ifstream ifs("../testcases/2_03.txt");
  ll n;
  ifs >> n;
  vector<pls> ps(n);
  for (ll i = 0; i < n; ++i) {
    char k;
    string s;
    ifs >> k >> s;
    ps[i] = {k, s};
  }

  run1(del_md_node1, "del_md_node1", ps);
  run2(del_md_node2, "del_md_node2", ps);
}